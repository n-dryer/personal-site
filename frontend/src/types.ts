import { LucideIcon } from 'lucide-react';

export interface Project {
  name: string;
  description: string;
}

export interface Experience {
  id: string;
  title: string;
  company: string;
  location: string;
  date: string;
  description: string;
  icon: LucideIcon;
  achievements: string[];
  technologies: string[];
  projects: Project[];
}

export interface Skill {
  id: string;
  name: string;
  level: number;
  icon?: string;
}

export interface SocialLink {
  name: string;
  url: string;
}

export interface UserData {
  fullName: string;
  bioLine: string;
  photoUrl: string;
  email: string;
  phone: string;
  location: string;
  socialLinks: SocialLink[];
  resumeUrl: string;
}

// Legacy interfaces to maintain compatibility during migration
export interface LegacyExperience {
  title: string;
  company: string;
  location: string;
  dates: string;
  blurb: string[];
}
