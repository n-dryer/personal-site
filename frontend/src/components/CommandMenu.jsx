import React, { useState, useEffect, useRef } from 'react';
import { Command } from 'cmdk';
import { Download, Mail, Linkedin, Github, Moon, Sun, ChevronRight } from 'lucide-react';

const CommandMenu = ({ userData, toggleTheme, darkMode }) => {
  const [open, setOpen] = useState(false);
  const [inputValue, setInputValue] = useState('');
  const inputRef = useRef(null);
  const listRef = useRef(null);

  // Define command items
  const items = [
    { 
      label: "Download résumé", 
      icon: Download, 
      href: "/nathan_dryer_resume.pdf", 
      shortcut: "⏎", 
      primary: true 
    },
    { 
      label: "Email Nate", 
      icon: Mail, 
      href: `mailto:${userData.email}` 
    },
    { 
      label: "LinkedIn", 
      icon: Linkedin, 
      href: userData.socialLinks.find(link => link.name === "LinkedIn")?.url || "#" 
    },
    { 
      label: "GitHub", 
      icon: Github, 
      href: userData.socialLinks.find(link => link.name === "GitHub")?.url || "#" 
    },
    { 
      label: `Toggle ${darkMode ? 'light' : 'dark'} mode`, 
      icon: darkMode ? Sun : Moon, 
      action: "theme" 
    },
  ];

  // Filter items based on search input
  const filteredItems = inputValue === '' 
    ? items 
    : items.filter(item => 
        item.label.toLowerCase().includes(inputValue.toLowerCase())
      );

  // Handle keyboard shortcuts
  useEffect(() => {
    const handleKeyDown = (e) => {
      // Open command palette with Cmd+K or Ctrl+K
      if ((e.metaKey || e.ctrlKey) && e.key === 'k') {
        e.preventDefault();
        setOpen(true);
      }
    };

    window.addEventListener('keydown', handleKeyDown);
    return () => window.removeEventListener('keydown', handleKeyDown);
  }, []);

  // Auto-focus input on palette open only on desktop
  useEffect(() => {
    if (open && window.innerWidth >= 768) {
      setTimeout(() => {
        inputRef.current?.focus();
      }, 100);
    }
  }, [open]);

  // Handle email copy
  const handleEmailCopy = () => {
    navigator.clipboard.writeText(userData.email);
    // You could add a toast notification here
    setOpen(false);
  };

  return (
    <Command
      label="Site command palette"
      className="w-full max-w-full md:max-w-md mx-auto rounded-2xl bg-white/65 dark:bg-slate-900/50
                backdrop-blur-sm shadow-lg ring-1 ring-black/5 dark:ring-white/10
                focus-within:ring-blue-500 dark:focus-within:ring-blue-400 transition"
    >
      <Command.Input
        ref={inputRef}
        value={inputValue}
        onValueChange={setInputValue}
        placeholder="Type a command…"
        className="h-12 w-full bg-transparent px-4 text-base text-gray-800 dark:text-white outline-none"
        autoComplete="off"
      />
      <Command.List 
        ref={listRef}
        className="grid gap-1 p-2"
      >
        {filteredItems.map(({ label, icon: Icon, href, shortcut, primary, action }) => (
          <Command.Item 
            key={label} 
            className="rounded-xl focus:bg-slate-100 dark:focus:bg-white/5 outline-none"
            role="menuitem"
          >
            {href ? (
              <a
                href={href}
                className={`flex items-center gap-3 px-4 py-3 rounded-xl w-full min-h-[48px]
                          ${primary 
                            ? "bg-gradient-to-r from-blue-500 to-blue-600 text-white hover:from-blue-600 hover:to-blue-700" 
                            : "hover:bg-slate-100 dark:hover:bg-white/5"}`}
                download={href.includes('resume') ? "Nathan_Dryer_Resume.pdf" : undefined}
              >
                <div className={`flex items-center justify-center w-8 h-8 ${primary 
                  ? "bg-white/20 rounded-full" 
                  : "text-blue-500 dark:text-blue-400"}`}>
                  <Icon className="w-5 h-5 shrink-0" />
                </div>
                <span className="grow text-left">{label}</span>
                {shortcut && (
                  <kbd className="px-2 py-1 text-xs bg-black/5 dark:bg-white/10 rounded">
                    {shortcut}
                  </kbd>
                )}
                {!primary && !shortcut && <ChevronRight className="w-4 h-4 opacity-50" />}
              </a>
            ) : (
              <button
                onClick={() => {
                  if (action === "theme") {
                    toggleTheme();
                  }
                  setOpen(false);
                }}
                className="flex items-center gap-3 px-4 py-3 rounded-xl w-full min-h-[48px]
                          hover:bg-slate-100 dark:hover:bg-white/5"
              >
                <div className="flex items-center justify-center w-8 h-8 text-blue-500 dark:text-blue-400">
                  <Icon className="w-5 h-5 shrink-0" />
                </div>
                <span className="grow text-left text-gray-800 dark:text-white">{label}</span>
              </button>
            )}
          </Command.Item>
        ))}
      </Command.List>
    </Command>
  );
};

export default CommandMenu;
