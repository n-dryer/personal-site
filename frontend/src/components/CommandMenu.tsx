import React, { useState, useCallback } from 'react';
import { Command } from 'cmdk';
import { motion, AnimatePresence } from 'framer-motion';
import CopyToClipboard from 'react-copy-to-clipboard';
import { Search, Mail, Clock, Linkedin, Github, X, Check } from 'lucide-react';

interface CommandMenuProps {
  isOpen: boolean;
  setIsOpen: (open: boolean) => void;
}

interface CommandItem {
  id: string;
  name: string;
  icon: React.ReactNode;
  keywords: string;
  action: () => void;
  component?: React.ReactNode;
}

export const CommandMenu: React.FC<CommandMenuProps> = ({
  isOpen,
  setIsOpen,
}) => {
  const [search, setSearch] = useState<string>('');
  const [copied, setCopied] = useState<boolean>(false);

  // Handle copy notification - direct state updates for better UX
  const handleCopyEmail = useCallback((): void => {
    setCopied(true);
    setTimeout(() => {
      setCopied(false);
      setIsOpen(false);
    }, 1500);
  }, [setIsOpen]);

  // Scroll to section helper - direct modal close for better UX
  const scrollToSection = useCallback(
    (id: string): void => {
      const element = document.getElementById(id);
      if (element) {
        element.scrollIntoView({ behavior: 'smooth' });
        setIsOpen(false); // Direct state update for modal close
      }
    },
    [setIsOpen]
  );

  // Open link in new tab helper - direct modal close for better UX
  const openLink = useCallback(
    (url: string): void => {
      window.open(url, '_blank', 'noopener,noreferrer');
      setIsOpen(false); // Direct state update for modal close
    },
    [setIsOpen]
  );

  // Simplified commands array (removed resume and skills)
  const commands: CommandItem[] = [
    {
      id: 'email',
      name: 'Copy Email Address',
      icon: <Mail size={18} className='text-token-primary-500' />,
      keywords: 'contact mail reach',
      action: handleCopyEmail,
      component: (
        <CopyToClipboard
          text='nathan.dryer@example.com'
          onCopy={handleCopyEmail}
        >
          <div className='flex w-full items-center justify-between'>
            <div className='flex items-center gap-3'>
              <Mail size={18} className='text-token-primary-500' />
              <div className='flex flex-col'>
                <span className='font-medium'>Copy Email</span>
                <span className='text-xs text-token-tertiary'>
                  nathan.dryer@example.com
                </span>
              </div>
            </div>
          </div>
        </CopyToClipboard>
      ),
    },
    {
      id: 'linkedin',
      name: 'LinkedIn Profile',
      icon: <Linkedin size={18} className='text-token-primary-500' />,
      keywords: 'social network professional connect',
      action: () => openLink('https://linkedin.com/in/nathan-dryer'),
    },
    {
      id: 'github',
      name: 'GitHub Profile',
      icon: <Github size={18} className='text-token-primary-500' />,
      keywords: 'code repository projects source',
      action: () => openLink('https://github.com/nathan-dryer'),
    },
    {
      id: 'timeline',
      name: 'View Experience Timeline',
      icon: <Clock size={18} className='text-token-primary-500' />,
      keywords: 'experience history career work jobs',
      action: () => scrollToSection('timeline'),
    },
  ];

  return (
    <AnimatePresence>
      {isOpen && (
        <motion.div
          className='fixed inset-0 z-50 flex items-start justify-center bg-black/60 px-4 pt-[15vh] sm:pt-[20vh]'
          initial={{ opacity: 0 }}
          animate={{ opacity: 1 }}
          exit={{ opacity: 0 }}
          transition={{ duration: 0.2 }}
          onClick={() => setIsOpen(false)}
          role='dialog'
        >
          <motion.div
            className='margin-component-y relative w-full max-w-md overflow-hidden rounded-xl shadow-2xl bg-token-primary'
            initial={{ scale: 0.95, opacity: 0, y: -10 }}
            animate={{ scale: 1, opacity: 1, y: 0 }}
            exit={{ scale: 0.95, opacity: 0, y: -10 }}
            transition={{ duration: 0.2, ease: 'easeOut' }}
            onClick={e => e.stopPropagation()}
          >
            <Command className='w-full'>
              <div className='flex items-center border-b px-4 pb-3 pt-4 border-token-primary'>
                <Search className='mr-3 h-4 w-4 text-token-tertiary' />
                <Command.Input
                  value={search}
                  onValueChange={setSearch}
                  className='placeholder-token-tertiary w-full bg-transparent py-2 text-base font-medium text-token-primary focus:outline-none'
                  placeholder='Find contact info or navigate...'
                  autoFocus
                />
                <button
                  className='rounded-md p-1.5 transition-colors text-token-tertiary hover:bg-interactive-secondary'
                  onClick={() => setIsOpen(false)}
                >
                  <X size={16} />
                </button>
              </div>

              {copied && (
                <div className='border-b border-token-primary-100 bg-token-primary-50 px-4 py-3'>
                  <div className='flex items-center text-token-primary-700'>
                    <Check size={16} className='mr-2' />
                    <p className='text-sm font-medium'>
                      Email copied to clipboard
                    </p>
                  </div>
                </div>
              )}

              <Command.List className='max-h-[300px] overflow-y-auto py-3'>
                <Command.Empty className='py-8 text-center'>
                  <p className='font-medium text-token-tertiary'>
                    No matching commands found
                  </p>
                  <p className='mt-1 text-sm text-token-tertiary'>
                    Try a different search term
                  </p>
                </Command.Empty>

                {commands.map(command => (
                  <Command.Item
                    key={command.id}
                    value={`${command.name.toLowerCase()} ${command.keywords}`}
                    onSelect={() => command.action()}
                    className='mx-2 flex cursor-pointer items-center justify-between rounded-lg px-4 py-3 transition-all duration-200 text-token-secondary hover:bg-token-primary-50 hover:text-token-primary-700'
                  >
                    {command.component || (
                      <div className='flex items-center gap-3'>
                        {command.icon}
                        <span className='font-medium'>{command.name}</span>
                      </div>
                    )}
                  </Command.Item>
                ))}
              </Command.List>
            </Command>
          </motion.div>
        </motion.div>
      )}
    </AnimatePresence>
  );
};
