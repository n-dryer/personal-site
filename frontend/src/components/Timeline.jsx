import React, { useEffect, useRef } from 'react';
import { motion, useAnimation } from 'framer-motion';
import { useInView } from 'react-intersection-observer';

const TimelineItem = ({ experience, index, isLatest }) => {
  const controls = useAnimation();
  const [ref, inView] = useInView({
    threshold: 0.2,
    triggerOnce: false
  });

  useEffect(() => {
    if (inView) {
      controls.start({
        opacity: 1,
        y: 0,
        transition: {
          type: "spring",
          stiffness: 50,
          damping: 15,
          delay: index * 0.1,
        }
      });
    } else {
      controls.start({
        opacity: 0,
        y: 20,
      });
    }
  }, [inView, controls, index]);

  return (
    <motion.li 
      ref={ref}
      className="w-full md:w-[260px] flex-shrink-0 scroll-snap-align-start"
      initial={{ opacity: 0, y: 20 }}
      animate={controls}
    >
      <div className={`relative p-5 rounded-xl bg-white/50 dark:bg-slate-900/50 backdrop-blur-sm 
                      shadow-sm ring-1 ring-black/5 dark:ring-white/10 h-full`}>
        {/* Timeline dot - with gradient ring for latest item */}
        <div className={`absolute -left-3 md:left-1/2 md:-top-3 md:-translate-x-1/2 w-6 h-6 rounded-full 
                        flex items-center justify-center
                        ${isLatest 
                          ? 'bg-gradient-to-r from-blue-400 to-blue-600 p-[3px]' 
                          : 'border-2 border-blue-500 dark:border-blue-400'}`}>
          <div className={`rounded-full ${isLatest 
            ? 'bg-white dark:bg-slate-900 w-full h-full flex items-center justify-center' 
            : 'bg-blue-500 dark:bg-blue-400 w-2 h-2'}`}>
            {isLatest && (
              <div className="w-2 h-2 rounded-full bg-blue-500 dark:bg-blue-400 
                            animate-pulse"></div>
            )}
          </div>
        </div>
        
        {/* Card content */}
        <div className="ml-4 md:ml-0 md:mt-4">
          <div className="text-sm font-medium text-blue-500 dark:text-blue-400 mb-1">
            {experience.dates}
          </div>
          <h3 className="text-lg font-bold text-gray-900 dark:text-white mb-1">
            {experience.title}
          </h3>
          <h4 className="text-md font-medium text-blue-600 dark:text-blue-500 mb-1">
            {experience.company}
          </h4>
          <p className="text-sm text-gray-500 dark:text-gray-400 mb-3">
            {experience.location}
          </p>
          <ul className="space-y-2 text-sm text-gray-700 dark:text-gray-300">
            {experience.blurb.map((point, i) => (
              <li key={i} className="flex items-start">
                <span className="mr-2 mt-1 text-blue-500 dark:text-blue-400">â€¢</span>
                <span>{point}</span>
              </li>
            ))}
          </ul>
        </div>
      </div>
    </motion.li>
  );
};

const Timeline = ({ userData }) => {
  const containerRef = useRef(null);

  return (
    <section 
      id="shiplog" 
      className="py-16 bg-gray-50 dark:bg-gray-950 mt-20 md:mt-32"
    >
      <div className="container mx-auto px-4">
        <h2 className="text-2xl md:text-3xl font-bold text-center mb-12 text-gray-900 dark:text-white">
          Experience Timeline
        </h2>
        
        <ul 
          ref={containerRef}
          className="shiplog relative flex flex-col md:flex-row gap-8 md:gap-6
                    h-[80vh] md:h-auto overflow-y-auto md:overflow-x-auto md:pb-8
                    scroll-snap-type-y md:scroll-snap-type-x scroll-snap-mandatory
                    pl-6 md:pl-0"
        >
          {/* Vertical line for mobile only */}
          <div className="absolute left-3 top-0 bottom-0 w-px bg-gradient-to-b from-blue-500 to-blue-600 md:hidden"></div>
          
          {userData.experiences.map((experience, index) => (
            <TimelineItem 
              key={index} 
              experience={experience} 
              index={index}
              isLatest={index === 0}
            />
          ))}
        </ul>
      </div>
    </section>
  );
};

export default Timeline;
