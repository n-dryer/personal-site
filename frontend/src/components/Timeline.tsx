import React from 'react';
import { VerticalTimeline, VerticalTimelineElement } from 'react-vertical-timeline-component';
import 'react-vertical-timeline-component/style.min.css';
import { UserData, Experience } from '../types';
import { useInView } from 'react-intersection-observer';
import { motion } from 'framer-motion';

interface TimelineProps {
  userData: UserData;
}

const Timeline: React.FC<TimelineProps> = ({ userData }) => {
  // Set up intersection observer to trigger animations
  const [ref, inView] = useInView({
    threshold: 0.1,
    triggerOnce: true,
  });

  // Calculate years of experience
  const currentYear = new Date().getFullYear();
  const startYear = 2009; // First job in the resume
  const yearsOfExperience = currentYear - startYear;

  return (
    <section id="timeline" ref={ref} className="py-16 bg-gray-50">
      <div className="container mx-auto px-4">
        <motion.div 
          className="text-center mb-16"
          initial={{ opacity: 0, y: 20 }}
          animate={inView ? { opacity: 1, y: 0 } : { opacity: 0, y: 20 }}
          transition={{ duration: 0.5 }}
        >
          <h2 className="text-3xl md:text-4xl font-bold text-gray-900 mb-4">Professional Journey</h2>
          <p className="text-lg text-gray-600 max-w-2xl mx-auto">
            With over {yearsOfExperience} years of experience, I've been building products that people love to use and buy.
          </p>
        </motion.div>

        <VerticalTimeline>
          {userData.experiences.map((experience: Experience, index: number) => (
            <VerticalTimelineElement
              key={index}
              className="vertical-timeline-element"
              contentStyle={{ background: 'white', color: '#333' }}
              contentArrowStyle={{ borderRight: '7px solid  white' }}
              date={experience.dates}
              iconStyle={{ background: 'rgb(79, 70, 229)', color: '#fff' }}
              icon={
                <div className="flex items-center justify-center h-full">
                  <svg xmlns="http://www.w3.org/2000/svg" className="h-6 w-6 text-white" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                    <path strokeLinecap="round" strokeLinejoin="round" strokeWidth={2} d="M21 13.255A23.931 23.931 0 0112 15c-3.183 0-6.22-.62-9-1.745M16 6V4a2 2 0 00-2-2h-4a2 2 0 00-2 2v2m4 6h.01M5 20h14a2 2 0 002-2V8a2 2 0 00-2-2H5a2 2 0 00-2 2v10a2 2 0 002 2z" />
                  </svg>
                </div>
              }
            >
              <h3 className="vertical-timeline-element-title text-xl font-bold">{experience.title}</h3>
              <h4 className="vertical-timeline-element-subtitle text-lg mt-1 text-primary-600">{experience.company}</h4>
              <p className="text-sm text-gray-500 mt-1 mb-3">{experience.location}</p>
              <ul className="list-disc pl-5 text-gray-700 space-y-2">
                {experience.blurb.map((point, i) => (
                  <li key={i}>{point}</li>
                ))}
              </ul>
            </VerticalTimelineElement>
          ))}
        </VerticalTimeline>
      </div>
    </section>
  );
};

export default Timeline;
