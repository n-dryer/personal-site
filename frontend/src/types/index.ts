// Re-export from portfolio.ts for backward compatibility
export * from './portfolio';
