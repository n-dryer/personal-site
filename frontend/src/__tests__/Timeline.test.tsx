import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Timeline } from '../components/Timeline';
import { Experience } from '../types';
import { Brain, Briefcase } from 'lucide-react';

// Mock framer-motion
jest.mock('framer-motion', () => ({
  motion: {
    div: ({
      children,
      animate,
      transition,
      initial,
      layout,
      ...props
    }: React.PropsWithChildren<any>) => (
      <div
        data-testid='motion-div'
        animate={animate ? JSON.stringify(animate) : undefined}
        transition={transition ? JSON.stringify(transition) : undefined}
        initial={initial ? JSON.stringify(initial) : undefined}
        layout={layout ? JSON.stringify(layout) : undefined}
        {...props}
      >
        {children}
      </div>
    ),
  },
  AnimatePresence: ({ children }: React.PropsWithChildren<any>) => (
    <>{children}</>
  ),
}));

// No longer need to mock react-vertical-timeline-component since we're using custom implementation

// Mock lucide-react icons
jest.mock('lucide-react', () => ({
  ChevronDown: () => <div data-testid='chevron-down'>Down</div>,
  ChevronRight: () => <div data-testid='chevron-right'>Right</div>,
  // Create a mock icon component that matches LucideIcon interface
  Brain: Object.assign(() => <div data-testid='brain-icon'>Brain</div>, {
    $$typeof: Symbol.for('react.forward_ref'),
  }),
  Briefcase: Object.assign(
    () => <div data-testid='briefcase-icon'>Briefcase</div>,
    { $$typeof: Symbol.for('react.forward_ref') }
  ),
}));

// Mock window.matchMedia
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

describe('Timeline Component', () => {
  // Create minimal mock data
  const mockExperienceData: Experience[] = [
    {
      id: 'exp-1',
      title: 'Test Position',
      company: 'Test Company',
      location: 'Test Location',
      date: '2020 - Present',
      description: 'Test description',
      icon: Brain,
      achievements: ['Achievement 1', 'Achievement 2'],
      technologies: ['Tech 1', 'Tech 2'],
      projects: [{ name: 'Project 1', description: 'Project 1 description' }],
    },
    {
      id: 'exp-2',
      title: 'Previous Position',
      company: 'Previous Company',
      location: 'Previous Location',
      date: '2018 - 2020',
      description: 'Previous role description',
      icon: Briefcase,
      achievements: ['Achievement A', 'Achievement B'],
      technologies: ['Tech A', 'Tech B'],
      projects: [{ name: 'Project A', description: 'Project A description' }],
    },
  ];

  it('renders correctly with collapsed items', () => {
    const { container } = render(
      <Timeline experienceData={mockExperienceData} />
    );
    expect(container).toMatchSnapshot();
  });

  it('expands an item when timeline icon is clicked', () => {
    render(<Timeline experienceData={mockExperienceData} />);

    // Find the first timeline icon and click it
    const timelineIcons = screen.getAllByRole('button');
    fireEvent.click(timelineIcons[0]);

    // Check that detailed information is now visible in expanded state
    expect(screen.getByText('Test Company')).toBeInTheDocument();
    // Year and location are now separated in expanded state
    expect(screen.getByText('2020-Present')).toBeInTheDocument();
    expect(screen.getByText('Test Location')).toBeInTheDocument();

    // Check that technologies are displayed
    expect(screen.getByText('Tech 1')).toBeInTheDocument();
    expect(screen.getByText('Tech 2')).toBeInTheDocument();

    // Check that the icon has expanded state
    expect(timelineIcons[0]).toHaveAttribute('aria-expanded', 'true');
  });

  it('collapses an expanded item when timeline icon is clicked again', () => {
    render(<Timeline experienceData={mockExperienceData} />);

    // Find the first timeline icon and click it twice
    const timelineIcons = screen.getAllByRole('button');
    fireEvent.click(timelineIcons[0]);
    fireEvent.click(timelineIcons[0]);

    // Check that expanded-only content is no longer visible (location should not be visible when collapsed)
    expect(screen.queryByText('Test Location')).not.toBeInTheDocument();
    // But company name should still be visible in collapsed state
    expect(screen.getByText('Test Company')).toBeInTheDocument();

    // Check that the icon has collapsed state
    expect(timelineIcons[0]).toHaveAttribute('aria-expanded', 'false');
  });

  it('displays proper visual hierarchy in collapsed state', () => {
    render(<Timeline experienceData={mockExperienceData} />);

    // Check that job titles are always visible (Primary Typography)
    expect(screen.getByText('Test Position')).toBeInTheDocument();
    expect(screen.getByText('Previous Position')).toBeInTheDocument();

    // Check that company names are now always visible (Secondary Typography)
    expect(screen.getByText('Test Company')).toBeInTheDocument();
    expect(screen.getByText('Previous Company')).toBeInTheDocument();

    // Check that year ranges are extracted and displayed (Tertiary Typography)
    expect(screen.getByText('2020-Present')).toBeInTheDocument();
    expect(screen.getByText('2018-2020')).toBeInTheDocument();

    // Check that expanded-only content is NOT visible in collapsed state
    expect(screen.queryByText('Test Location')).not.toBeInTheDocument();
    expect(screen.queryByText('Test description')).not.toBeInTheDocument();
  });

  it('only allows one expanded item at a time', () => {
    render(<Timeline experienceData={mockExperienceData} />);

    // Get timeline icon buttons specifically (they have the timeline-icon class)
    const timelineIcons = screen
      .getAllByRole('button')
      .filter(button => button.classList.contains('timeline-icon'));

    // Expand first item
    fireEvent.click(timelineIcons[0]);
    expect(timelineIcons[0]).toHaveAttribute('aria-expanded', 'true');
    expect(timelineIcons[1]).toHaveAttribute('aria-expanded', 'false');

    // Expand second item - should collapse first
    fireEvent.click(timelineIcons[1]);
    expect(timelineIcons[0]).toHaveAttribute('aria-expanded', 'false');
    expect(timelineIcons[1]).toHaveAttribute('aria-expanded', 'true');
  });

  it('displays alternating left-right layout correctly', () => {
    const { container } = render(
      <Timeline experienceData={mockExperienceData} />
    );

    // Check that first item (index 0, even position) has timeline-left class
    const firstTimelineItem = container.querySelector('.timeline-left');
    expect(firstTimelineItem).toBeInTheDocument();

    // Check that second item (index 1, odd position) has timeline-right class
    const secondTimelineItem = container.querySelector('.timeline-right');
    expect(secondTimelineItem).toBeInTheDocument();

    // Check that timeline grid items have proper CSS Grid classes
    const timelineGridItems = container.querySelectorAll(
      'div[class*="timeline-grid-item"]'
    );
    expect(timelineGridItems.length).toBe(2); // Should have 2 timeline items

    // Check that year labels are positioned above cards
    const yearLabelsAbove = container.querySelectorAll(
      '.timeline-year-label-above'
    );
    expect(yearLabelsAbove.length).toBe(2); // Should have 2 year labels above cards

    // Check that year labels are positioned independently (three-element architecture)
    const yearLabels = container.querySelectorAll(
      'div[class*="timeline-year-label"]'
    );
    expect(yearLabels.length).toBe(2); // Should have 2 year labels

    // Verify that all cards have consistent text-left alignment (no more text-right)
    const allCardContents = container.querySelectorAll('div.timeline-card');
    allCardContents.forEach(card => {
      expect(card).toHaveClass('text-left');
      expect(card).not.toHaveClass('text-right');
    });
  });

  it('displays clean cards without chevron icons', () => {
    render(<Timeline experienceData={mockExperienceData} />);

    // Check that chevron icons are NOT present (removed for cleaner design)
    const chevronIcons = screen.queryAllByTestId('chevron-right');
    expect(chevronIcons.length).toBe(0); // No chevron icons should be present

    // Check that old text-based hints are removed
    expect(
      screen.queryByText('Click the timeline icon to collapse')
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText('Click on any timeline node or card to expand details')
    ).not.toBeInTheDocument();
  });

  it('has proper hover effects and accessibility for modern indicators', () => {
    const { container } = render(
      <Timeline experienceData={mockExperienceData} />
    );

    // Check that cards have cursor-pointer class for hover indication
    const timelineCards = container.querySelectorAll(
      'div[class*="cursor-pointer"]'
    );
    expect(timelineCards.length).toBeGreaterThan(0);

    // Check that timeline icons have proper ARIA labels with expand/collapse state
    const timelineIcons = screen.getAllByRole('button');
    expect(timelineIcons[0]).toHaveAttribute('aria-expanded', 'false');
    expect(timelineIcons[0]).toHaveAttribute(
      'aria-label',
      expect.stringContaining('Expand details')
    );
  });
});
