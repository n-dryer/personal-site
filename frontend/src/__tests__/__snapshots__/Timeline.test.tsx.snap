// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Component renders correctly with collapsed items 1`] = `
<div>
  <section
    class="bg-canvas py-[var(--space-section)] dark:bg-canvas-dark"
    id="timeline"
  >
    <div
      class="container mx-auto px-4"
    >
      <h2
        class="text-high-contrast mb-12 text-center text-3xl font-bold"
      >
        Professional Timeline
      </h2>
      <div
        class="timeline-grid-container relative mx-auto w-full max-w-none px-4 md:max-w-6xl md:px-8"
      >
        <div
          class="timeline-line-desktop hidden w-0.5 bg-gradient-to-b from-primary-400 via-primary-500 to-primary-600 md:block"
          style="height: 96px; margin-top: 46px;"
        />
        <div
          class="absolute w-0.5 bg-gradient-to-b from-primary-400 via-primary-500 to-primary-600 md:hidden"
          style="height: 96px; top: 46px; left: 22px;"
        />
        <div
          class="timeline-grid-item relative timeline-left"
          style="min-height: 96px; padding-top: 24px; padding-bottom: 16px; margin-bottom: 16px;"
        >
          <button
            aria-controls="timeline-content-exp-1"
            aria-describedby="timeline-content-exp-1"
            aria-expanded="false"
            aria-label="Expand details for Test Position at Test Company"
            aria-labelledby="timeline-title-exp-1"
            class="timeline-icon timeline-icon-focus flex h-11 min-h-[44px] w-11 min-w-[44px] cursor-pointer items-center justify-center rounded-full bg-primary-500 text-white shadow-lg hover:scale-110 hover:bg-primary-600 hover:shadow-xl active:scale-95 animate-timeline-pulse transition-all duration-300"
            role="button"
            type="button"
          >
            <div
              data-testid="brain-icon"
            >
              Brain
            </div>
          </button>
          <div
            animate="{\\"transition\\":{\\"duration\\":0.4,\\"ease\\":[0.4,0,0.2,1],\\"type\\":\\"spring\\",\\"stiffness\\":300,\\"damping\\":30}}"
            class="timeline-card-wrapper relative w-full md:w-auto z-timeline-card md:max-w-sm lg:max-w-md xl:max-w-lg "
            data-testid="motion-div"
            id="timeline-content-exp-1"
            layout="\\"position\\""
            style="margin-top: 0px;"
          >
            <div
              class="timeline-year-label-above"
            >
              <p
                class="whitespace-nowrap text-lg font-semibold text-primary-600 transition-all duration-300 dark:text-primary-400 md:text-xl"
              >
                2020-Present
              </p>
            </div>
            <div
              aria-controls="timeline-expanded-exp-1"
              aria-labelledby="timeline-title-exp-1"
              class="timeline-card rounded-lg bg-white/95 shadow-md backdrop-blur-sm dark:bg-slate-800/95 timeline-card-hover transition-all duration-300 text-left p-4 timeline-card-focus relative cursor-pointer"
              data-testid="motion-div"
              data-view-transition-name="timeline-card-1"
              layout="\\"position\\""
              role="button"
              style="--view-transition-name: timeline-card-1;"
              tabindex="0"
            >
              <div
                data-testid="motion-div"
                layout="\\"position\\""
              >
                <h3
                  class="mb-2 text-lg text-high-contrast font-semibold leading-tight"
                  id="timeline-title-exp-1"
                >
                  Test Company
                </h3>
                <p
                  class="mb-2 text-base text-medium-contrast font-medium leading-snug"
                >
                  Test Position
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="timeline-grid-item relative timeline-right"
          style="min-height: 96px; padding-top: 24px; padding-bottom: 16px; margin-bottom: 16px;"
        >
          <button
            aria-controls="timeline-content-exp-2"
            aria-describedby="timeline-content-exp-2"
            aria-expanded="false"
            aria-label="Expand details for Previous Position at Previous Company"
            aria-labelledby="timeline-title-exp-2"
            class="timeline-icon timeline-icon-focus flex h-11 min-h-[44px] w-11 min-w-[44px] cursor-pointer items-center justify-center rounded-full bg-primary-500 text-white shadow-lg hover:scale-110 hover:bg-primary-600 hover:shadow-xl active:scale-95 animate-timeline-pulse transition-all duration-300"
            role="button"
            type="button"
          >
            <div
              data-testid="briefcase-icon"
            >
              Briefcase
            </div>
          </button>
          <div
            animate="{\\"transition\\":{\\"duration\\":0.4,\\"ease\\":[0.4,0,0.2,1],\\"type\\":\\"spring\\",\\"stiffness\\":300,\\"damping\\":30}}"
            class="timeline-card-wrapper relative w-full md:w-auto z-timeline-card md:max-w-sm lg:max-w-md xl:max-w-lg "
            data-testid="motion-div"
            id="timeline-content-exp-2"
            layout="\\"position\\""
            style="margin-top: 0px;"
          >
            <div
              class="timeline-year-label-above"
            >
              <p
                class="whitespace-nowrap text-lg font-semibold text-primary-600 transition-all duration-300 dark:text-primary-400 md:text-xl"
              >
                2018-2020
              </p>
            </div>
            <div
              aria-controls="timeline-expanded-exp-2"
              aria-labelledby="timeline-title-exp-2"
              class="timeline-card rounded-lg bg-white/95 shadow-md backdrop-blur-sm dark:bg-slate-800/95 timeline-card-hover transition-all duration-300 text-left p-4 timeline-card-focus relative cursor-pointer"
              data-testid="motion-div"
              data-view-transition-name="timeline-card-2"
              layout="\\"position\\""
              role="button"
              style="--view-transition-name: timeline-card-2;"
              tabindex="0"
            >
              <div
                data-testid="motion-div"
                layout="\\"position\\""
              >
                <h3
                  class="mb-2 text-lg text-high-contrast font-semibold leading-tight"
                  id="timeline-title-exp-2"
                >
                  Previous Company
                </h3>
                <p
                  class="mb-2 text-base text-medium-contrast font-medium leading-snug"
                >
                  Previous Position
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
